tools/prepare_dataset.py

```diff
@@ -0,0 +1,224 @@
+import json
+import os
+from functools import cached_property
+from multiprocessing import Pool
+from pathlib import Path
+
+import numpy as np
+from datasets import load_dataset, load_from_disk
+from torch import distributed as dist
+from tqdm import tqdm
+from transformers import AutoTokenizer, logging
+
+from fast_llm.config import Field, FieldHint, check_field, config_class
+from fast_llm.data.mmap import MMapIndexedDataset
+from fast_llm.engine.config_utils.data_type import DataType
+from fast_llm.engine.config_utils.runnable import RunnableConfig
+from fast_llm.utils import Assert
+
+logging.set_verbosity_error()
+
+
+@config_class
+class PrepareDatasetConfig(RunnableConfig):
+    dataset_name_or_path: str = Field(
+        desc="Name or path of the dataset.",
+        hint=FieldHint.core,
+    )
+    tokenizer_path_or_name: str = Field(
+        desc="Path or name of the tokenizer.",
+        hint=FieldHint.core,
+    )
+    output_dir: str = Field(
+        desc="Output directory for the processed dataset.",
+        hint=FieldHint.core,
+    )
+    num_processes_load: int = Field(
+        default=1,
+        desc="Number of workers in load_dataset() call.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.geq, 0),
+    )
+    num_processes_map: int = Field(
+        default=1,
+        desc="Number of workers in .map() call.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.geq, 0),
+    )
+    num_processes_save: int = Field(
+        default=1,
+        desc="Number of processes for saving the mmap'ed datasets.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.geq, 0),
+    )
+    num_tokens_per_shard: int = Field(
+        default=1000000000,
+        desc="Approximate number of tokens per shard.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.geq, 1),
+    )
+    dataset_config_name: None | str = Field(
+        default=None,
+        desc="Specific configuration name for the dataset.",
+        hint=FieldHint.optional,
+    )
+    dataset_split: str = Field(
```

Comment:
Consider making a sub-config for input dataset to avoid all the `dataset` prefixes?

Reply:
Also what does the split do?

Reply:
hf datasets come with multiple splits. we need to select one, and that is usually `train`.